Added Euro functionality for Euro trading pairs*

Leunam opened the pull request:
* Tested on Binance!

- user_config.yaml
Added is_euro bool variable, False by default.
- corecito_account.py
Added 'self.is_euro' var to store is_euro config variable
- corecito.py
Added conditions to trade in Euro if var 'euro' is True

Leunam committed:
Added Euro functionality for Euro trading pairs*

* Tested on Binance!

sebaslogen commented:
Looking at the Euro pairs they don't seem to have anything special. The base currency in most crypto pairs is Bitcoin and the base currency in all Euro and USD pairs is the fiat, so in theory it should work just adjusting the config. Do you know if this is the case?

I tried myself but the timestamp prevents me from testing this, could you please check if the code from main branch works using this config below?
```
binance_trading_pair: BTCEUR # This must be one of the pairs listed in Binance
binance_core_number_currency: BTC # # This must be one of the coins listed in Binance
binance_base_currency: EUR # # This must be one of the coins listed in Binance
binance_max_decimals_buy: 6
binance_max_decimals_sell: 6
```

sebaslogen commented:
Maybe the config above works but it's not want you want. 😅 

I built the first version of corecito to generate ETH benefits using BTC as the base currency and my assumption now is that if you change just the settings to use the BTCEUR pair, like I described above, it will generate BTC benefits based on a core number of BTC.

What users probably want instead is to generate EUR/fiat benefits based on a core number of EUR/fiat. 
In this case, indeed most ecuations have to be inverted and probably the right name for the boolean config variable should be something like `invert_pair` (instead of `is_euro`) and the config should look something like this:
```
binance_trading_pair: BTCEUR # This must be one of the pairs listed in Binance it was ETHBTC
binance_core_number_currency: EUR # The currency in which to produce benefits and denominate the core number. E.g. 1000 EUR or 10 ETH
binance_base_currency: EUR # The currency in which the transactions are made (usually the left currency in the pair)
binance_max_decimals_buy: 6
binance_max_decimals_sell: 6
```

Leunam commented:
```
binance_trading_pair: BTCEUR # This must be one of the pairs listed in Binance
binance_core_number_currency: BTC # # This must be one of the coins listed in Binance
binance_base_currency: EUR # # This must be one of the coins listed in Binance
binance_max_decimals_buy: 6
binance_max_decimals_sell: 6
```
This configuration will raise the next issue:
1 -i.e. We have 0.3 bitcoin to play with and 10000€ in the fiat wallet
 As we use the BTC/EUR pair, the buy and sell ticker are gonna be read in euros, as Binance returns that.
i.e. 
`buy_price `= 30000
`sell_price `= 30100

2 - the `deviated_core_number `will result as
`deviated_core_number = balances['base_currency_available'] / buy_price`
deviated_core_number = 10000 / 30000
deviated_core_number = 0.33

Corecito will think that the price has plummeted printing that the price drop 100% as:

![image](https://user-images.githubusercontent.com/576484/104257563-074c8580-547e-11eb-8cc2-b55c89bc9e7b.png)

I used this configuration in my first run. I wanted to see what would do your code in a different Pair.
That's why I had to multiply the base currency instead of using the division, but! Using BTC as core_number_currency will bring the next issue:
`deviated_core_number = balances['base_currency_available'] / buy_price`
`deviated_core_number `= 10000 * 30000
`deviated_core_number `= 300000000
So we get the same issue as before, but this time the price exploded.

![image](https://user-images.githubusercontent.com/576484/104257479-d9674100-547d-11eb-95d9-edb3f46265a8.png)

Conclusion is that the core_number_currency in fiat pairs has to be the fiat symbol code (EUR, USDT, etc...), forcing us to create the fiat conditional statements.

====================

> What users probably want instead is to generate EUR/fiat benefits based on a core number of EUR/fiat.

**Correct**.

> In this case, indeed most ecuations have to be inverted and probably the right name for the boolean config variable should be something like `invert_pair` (instead of `is_euro`) and the config should look something like this:

I agree, the var name is not accurate, as it could be USDT, GBP, DAI, PAX, etc...
@sebaslogen  what about `is_fiat` or just `fiat`?

> 
> ```
> binance_trading_pair: BTCEUR # This must be one of the pairs listed in Binance it was ETHBTC
> binance_core_number_currency: EUR # The currency in which to produce benefits and denominate the core number. E.g. 1000 EUR or 10 ETH
> binance_base_currency: EUR # The currency in which the transactions are made (usually the left currency in the pair)
> binance_max_decimals_buy: 6
> binance_max_decimals_sell: 6
> ```

This configuration would be like using a EUR/EUR pair, bringing back the same issue as reported above.

![image](https://user-images.githubusercontent.com/576484/104258392-c9e8f780-547f-11eb-9312-225a60ba176a.png)

sebaslogen commented:
The rename to `is_fiat` is fine and the code also looks good to me.
I was hoping to generalize your code even further for all pairs where you want to obtain profits on the other side of the pair, but that's too ambitious and not needed at all for now.

Can you rename the variable and I press merge?

Leunam committed:
[Code] - Core refactor, var is_euro -> is_fiat

Leunam commented:
> Can you rename the variable and I press merge?

Done! 
Please be sure to run some Tests when running your BTC/ETH pair!!!

sebaslogen commented on config/default_config.yaml:
```diff
@@ -26,9 +26,20 @@ binance_base_currency: BTC # # This must be one of the coins listed in Binance
 binance_max_decimals_sell: 6
 
+# FIAT SETTINGS
+is_fiat: False  # TODO: Not sure if the Euro check should be here, it can be in corecito.py code just checking the <exchange>_trading_pair string
```
The documentation here is old and can be removed.
I think adding a documentation example on this part of the file about how to config BTC/EUR fiat trading for Binance will help a lot new users.

Leunam replied:
Updated!

sebaslogen commented on corecito.py:
```diff
@@ -33,14 +35,26 @@ async def main():
+      
+      #Adjust logger.info text format if 'fiat'
+      if fiat:
+        logger.info(f'Core number: €{account.core_number}')
```
Why is the `account.core_number_currency` not shown for fiat? Is the core_number_currency in this case not EUR?

Leunam replied:
This is just styling the text format to replace 'EUR' for the euro symbol €. But you are right, as the code should be generic for any fiat or stable coin I will update this.

Leunam replied:
Updated!

sebaslogen commented on corecito.py:
```diff
@@ -50,20 +64,45 @@ async def main():
         if (not config['safe_mode_on']):
-          await account.order_market_buy(tx_result, excess)
+          if fiat:
+            await account.order_market_sell(tx_result)
```
If we make a sell order for fiat, shouldn't we also use the sell_price to calculate the excess, instead of the buy_price?

Leunam replied:
The `buy_price` is the closest price to the last known price a bider is willing to pay. If we use Limit Orders, then we can use the `sell_price`, but as we are using Market Orders, the `buy_price` should be used.

Umm, now that you mention this, actually we should use a new `last_price` variable to calculate it? Due to the Market Orders nature.

sebaslogen replied:
I think the last_price means the last transaction price, but that doesn't mean there is still an order open in the orderbook with that price anymore.

The buy and sell prices are the top of the orderbook from both sides. So a market order should always start by buying/selling the top of the order book and if the market order is bigger than what's available in the top order, then the next order will be consumed.

sebaslogen replied:
Regarding the buy/sell price I describe the scenario with both pairs:
In ETH/BTC, coreNumberIncreased means the value of my BTCs in ETHs is higher than my core number so I have to buy some ETHs with my BTCs to reduce my amount of BTCs.
To buy some ETHs we have to look at the **sell_price**.

In BTC/EUR, coreNumberIncreased means the value of my BTCs in EURs is higher than my core number so I have to sell some BTCs to take down the EURs value of my BTC.
To sell some BTCs we have to look at the **buy_price**.

Can you please check if this is correct?
If this is the case, then the existing logic was incorrect and the buy/sell prices should be inverted for the non-fiat transactions.

Leunam replied:
Exactly, if you execute `account.order_market_buy` you should take `sell_price` for calculations, as you buy from the sellers (ask)ing for the minimum price they are willing to accept.
So, in `coreNumberIncreased` it should be:
```
        if fiat:
          tx_result = round(excess / buy_price, account.max_decimals_buy)
        else:
          tx_result = round(excess * sell_price, account.max_decimals_buy)
```
because we execute:
```
        if (not config['safe_mode_on']):
          if fiat:
            await account.order_market_sell(tx_result) <---- as we sell, we calculate with buy_price for fiat
          else:
            await account.order_market_buy(tx_result, excess) <--- as we buy, we calculate with sell_price for crypto
```

sebaslogen replied:
Perfect, double better your PR, it helped find bugs in the existing script and it was still making profits 😅 😅 😅 
Would you mind making the change/fix in this PR?

sebaslogen commented on corecito.py:
```diff
@@ -50,20 +64,45 @@ async def main():
+        #Check is fiat is True to adjust messages format and tx_result var and 'missing' has to be divided by the sell_price
+        if fiat:
+          logger.info(f'Decreased {decrease_percentage:.2f}% - missing {missing:.6f} {account.core_number_currency} denominated in {account.base_currency}')
+          tx_result = round(missing / sell_price, account.max_decimals_sell)
```
Same question as above inverted, shouldn't we use here the buy_price for fiat?

Leunam replied:
Same as before, as we are using Market Orders, you buy to the next Seller (Ask) in line in the Order Book.

sebaslogen reviewed (changes requested):
Awesome feature! This will be really useful for corecito users
![](https://media2.giphy.com/media/3ohzdIuqJoo8QdKlnW/giphy-downsized-medium.gif)


A few small comments and a question about the inversion of buy/sell prices.

Leunam committed:
Code refactor

Visual Studio python project files add to gitignore.
default_config.yaml refactor

Leunam committed:
Refactor - € symbol removed from loggers

As we can use any fiat or stable coin, this doesn't make sense anymore. Removed!

Leunam committed:
Update .gitignore

Leunam committed:
Merge branch 'main' into eurocore

sebaslogen commented on corecito.py:
```diff
@@ -54,19 +62,37 @@ async def main():
+            await account.order_market_buy(tx_result, excess)
 
       elif coreNumberDecreased(account.core_number, deviated_core_number, account.min_core_number_decrease_percentage, account.max_core_number_decrease_percentage):
-        logger.logCoreNumberDecreased(decrease_percentage, missing, account.core_number_currency, account.base_currency, telegram)
-        tx_result = missing * sell_price
-        logger.logBuyMissing(tx_result, account.base_currency, buy_price, missing, account.core_number_currency, telegram)
```
It seems the merge deleted the 2 logger lines that should replace the old logger.info lines below.

Leunam replied:
ummm, I'm on it!

Leunam committed:
Logger merge bug fixed

Leunam committed:
buy_price, sell_price vars reversed

sebaslogen commented on corecito.py:
```diff
@@ -54,19 +62,38 @@ async def main():
+          tx_result = round(excess / buy_price, account.max_decimals_buy)
+        else:
+          tx_result = round(excess * sell_price, account.max_decimals_buy)
         logger.logSellExcess(tx_result, account.base_currency, buy_price, excess, account.core_number_currency, telegram)
```
I just realize that the logged price here(buy_price) and in `coreNumberDecreased` are incorrect for fiat, because they use the other side of the order book. Maybe it's better to extract the price for the transaction and the log to a `tx_price` variable.
What do you think?

Leunam replied:
yeah, it makes sense, nice catch, I will update it right now.

Leunam replied:
Done!

Leunam committed:
new var 'tx_price' to log buys/sells for any pair
